crates/wash-lib/src/plugin/subcommand.rs

```diff
@@ -171,18 +186,79 @@ impl SubcommandRunner {
         let vars: Vec<_> = std::env::vars()
             .filter(|(k, _)| k.starts_with(&env_prefix))
             .collect();
-        plugin.store.data_mut().ctx = WasiCtxBuilder::new()
-            // Disable socket connections for now. We may gradually open this up later
-            .socket_addr_check(|_, _| false)
-            .inherit_stderr()
-            .inherit_stdin()
+        let mut ctx = WasiCtxBuilder::new();
+        for dir in dirs {
+            // To avoid relative dirs and permissions issues, we canonicalize the host path
+            let canonicalized = tokio::fs::canonicalize(&dir.host_path)
+                .await
+                .context("Error when canonicalizing given path")?;
+            // We need this later and will have to return an error anyway if this fails
+            let str_canonical = canonicalized.to_str().ok_or_else(|| anyhow::anyhow!("Canonicalized path cannot be converted to a string for use in a plugin. This is a limitation of the WASI API"))?.to_string();
+            // Check if the path is a file or a dir so we can handle permissions accordingly
+            let is_dir = tokio::fs::metadata(&canonicalized)
+                .await
+                .map(|m| m.is_dir())
+                .context("Error when checking if path is a file or a dir")?;
+            let (host_path, guest_path, dir_perms) = match (is_dir, dir.component_path) {
+                (true, Some(path)) => (canonicalized, path, DIRECTORY_ALLOW),
+                (false, Some(path)) => (
+                    canonicalized
+                        .parent()
+                        .ok_or_else(|| anyhow::anyhow!("Could not get parent of given file"))?
+                        .to_path_buf(),
+                    path,
+                    DIRECTORY_DENY,
+                ),
+                (true, None) => (
+                    canonicalized.clone(),
+                    str_canonical.clone(),
+                    DIRECTORY_ALLOW,
+                ),
+                (false, None) => {
+                    let parent = canonicalized
+                        .parent()
+                        .ok_or_else(|| anyhow::anyhow!("Could not get parent of given file"))?
+                        .to_path_buf();
+                    (
+                        parent.clone(),
+                        // SAFETY: We already checked that canonicalized was a string above so we
+                        // can just unwrap here
+                        parent.to_str().unwrap().to_string(),
+                        DIRECTORY_DENY,
+                    )
+                }
+            };
+            println!(
+                "Host path: {}, guest_path: {}",
+                host_path.display(),
+                guest_path
+            );
```

Comment:
Should this be a log to avoid interfering with output? Or at least to stderr?

Reply:
This was just here for debugging purposes. Will be gone shortly